docker_push_latest_if_changed.py

```diff
@@ -1,5 +1,180 @@
 #!/usr/bin/env python3
 import argparse
+import hashlib
+import json
+import subprocess
+import typing
+
+
+class ImageNotFoundError(ValueError):
+    pass
+
+
+class _ImageKey(typing.NamedTuple):
+    commands_hash: str
+    packages_hash: str
+
+
+def _docker_push_latest_if_changed(source, target, *, is_dry_run):
+    _validate_source(source)
+    target = _get_sanitized_target(source, target)
+
+    print('Pushing source image')
+    _push_image(source, is_dry_run=is_dry_run)
+    try:
+        print('Pulling target image...')
+        _pull_image(target)
+    except ImageNotFoundError:
+        print((
+            f'Target image {target} was not found in the registry. '
+            'Going to attempt to tag and push the target image anyway.'
+        ))
```

Comment:
the parens here do nothing

Reply:
same for the rest of the files, I'm not going to point out each one but hope that you fix each of them 👍

Reply:
You're right, they do nothing!